File: app/schemas/category.py
from pydantic import BaseModel, ConfigDict, Field


class CategorySpecificationDefinitionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    key: str
    label: str
    group: str
    data_type: str
    unit: str | None
    filter_type: str
    required: bool
    comparable: bool
    facetable: bool
    applicability: str = "optional"
    sort_order: int
    options: list[str] | None = None
    item_schema: dict | None = None


class CategoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    name: str
    slug: str
    parent_id: int | None
    priority: str = "P2"
    is_group: bool = False
    sort_order: int = 0
    enabled: bool = True
    product_count: int = 0
    total_products: int = 0
    spec_definitions: list[CategorySpecificationDefinitionRead] = Field(default_factory=list)
    children: list["CategoryRead"] = Field(default_factory=list)

File: frontend/types/category.ts
export interface CategorySpecDefinition {
  id: number;
  key: string;
  label: string;
  group: string;
  data_type: string;
  unit: string | null;
  filter_type: string;
  required: boolean;
  comparable: boolean;
  facetable: boolean;
  sort_order: number;
  options: string[] | null;
}

export interface CategoryRead {
  id: number;
  name: string;
  slug: string;
  parent_id: number | null;
  priority: string;
  is_group: boolean;
  sort_order: number;
  enabled: boolean;
  product_count: number;
  total_products: number;
  spec_definitions: CategorySpecDefinition[];
  children: CategoryRead[];
}

export function flattenCategories(tree: CategoryRead[]): CategoryRead[] {
  const result: CategoryRead[] = [];
  for (const node of tree) {
    result.push(node);
    result.push(...flattenCategories(node.children));
  }
  return result;
}

export function findCategoryBySlug(root: CategoryRead, slug: string): CategoryRead | null {
  if (root.slug === slug) return root;
  for (const child of root.children) {
    const match = findCategoryBySlug(child, slug);
    if (match) return match;
  }
  return null;
}

/** Ancestors (root first) ending at the matching node. */
export function findCategoryChain(tree: CategoryRead[], slug: string): CategoryRead[] | null {
  for (const node of tree) {
    const chain = findPath(node, slug);
    if (chain) return chain;
  }
  return null;
}

function findPath(node: CategoryRead, slug: string): CategoryRead[] | null {
  if (node.slug === slug) return [node];
  for (const child of node.children) {
    const childPath = findPath(child, slug);
    if (childPath) return [node, ...childPath];
  }
  return null;
}

/** Top-level groups displayed on the home explorer, skipping pure wrapper nodes. */
export function mainGroups(tree: CategoryRead[]): CategoryRead[] {
  const enabled = tree.filter((node) => node.enabled);
  for (const root of enabled) {
    if (
      root.is_group &&
      root.product_count === 0 &&
      root.children.some((child) => child.is_group || child.children.length > 0)
    ) {
      return root.children.filter((child) => child.enabled);
    }
  }
  return enabled;
}

export function isLeafCategory(category: CategoryRead): boolean {
  return category.children.length === 0;
}